src/textual/_collections.py

```diff
@@ -0,0 +1,65 @@
+"""Provides collection-based utility code."""
+
+from __future__ import annotations
+from typing import Generic, TypeVar, Iterator, overload, Iterable
+
+T = TypeVar("T")
+
+
+class ImmutableSequence(Generic[T]):
+    """Class to wrap a sequence of some sort, but not allow modification."""
+
+    def __init__(self, wrap: Iterable[T]) -> None:
+        """Initialise the immutable sequence.
+
+        Args:
+            wrap (Iterable[T]): The iterable value being wrapped.
+        """
+        self._list = list(wrap)
```

Comment:
wouldn't it be better to just hold a reference to the original sequence? Why copy it? We're not allowing modification anyway.

Reply:
(and, not to retract from the above: if copying, why not make it tuple() while we're at it, to REALLY disallow modification?)

Reply:
> wouldn't it be better to just hold a reference to the original sequence? Why copy it? We're not allowing modification anyway.

The request for this was, in part, that it be able to take any reasonable iterable while allowing it to be used list-like.

Reply:
I'll give it one more shot, then:

1. I was thinking more something like `KeysView` and `ItemsView` from Python's stdlib. They are "aware of perf tradeoffs", so to speak. So here we would have `SequenceView`, and use it where appropriate. There can be also `ImmutableSequence`, but it doesn't mean it has to be used in the particular case of TreeNode's children.
2. How about adding a check in ImmutableSequence's `__init__()`: if `wrap` is an instance of `Sequence` then just store a reference; other convert to tuple.

Apologies for "insisting", I promise this is the last of them, and hope you find the above useful.

Reply:
Nah, appreciate it. Of course unlike those views I want this to be a subscriptable object. But yes, you're right, makes more sense to only convert when absolutely needed.